README
Introduction:
=============

Loudmouth is a lightweight and easy-to-use C library for programming with the
Jabber protocol. It's designed to be easy to get started with and yet
extensible to let you do anything the Jabber protocol allows.

The goal is to fully support Linux, *BSD, Mac OSX and Windows. It has
currently only been tested on Linux (Red Hat 8.0, Debian Unstable) and FreeBSD
that I know of. If someone with access to other machines can provide
feedback I would be most grateful.


Requirements:
=============

Glib >= 2.12.4:
http://ftp.gnome.org/pub/GNOME/sources/glib/2.16/

gtk-doc (optional, if you want documentation built):
ftp://ftp.gnome.org/pub/GNOME/sources/gtk-doc/1.0

If you want our unit tests through the g_test framework:
Glib >= 2.16.0

If you want SSL-support you'll also need:
GnuTLS >= 1.2.0 with libtasn support.

Building:
=========

./configure
make
make install

SSL Support:
============
You can choose between using GnuTLS (recommended) and OpenSSL to enable SSL
support in Loudmouth. You choose which backend to use by passing the
--with-ssl= flag to configure. GnuTLS is the default and will be used unless
you pass the flag, use "openssl" for OpenSSL.

Debugging:
==========

You can get debug information printed to STDOUT by using the environment
variable LM_DEBUG.

export LM_DEBUG=<keyword>

Where keyword is one or more (separated by :) of the following:

* "NET": Shows all the data sent over the network.
* "PARSER": Shows debug output from the parser.
* "VERBOSE": Shows various verbose output.
* "ALL": Enables all output.


Questions, remarks, bug reports:
================================

This fork of loudmouth is lazily maintained by MCabber community, members
of which can be found in official mcabber chatroom: <xmpp:mcabber@conf.lilotux.net>.

New releases from the MCabber team are available at
<https://mcabber.com/files/loudmouth/>

Bugs can be also reported (to not lose track of them) to github bugtracker:
<https://github.com/mcabber/loudmouth/issues>


Official loudmouth project is no more maintained - both
<http://projects.imendio.com/loudmouth> and
<http://loudmouth-project.org>.


NEWS
** Unused, please look at the release notes **




Changes in 1.2.0:
---------------
* This is the same as 1.1.5 but marked as stable and released as 1.2.0.
* For a complete list of changes since 1.0.5 look at the changes in 1.1.1-1.1.5

Changes in 1.1.5:
-----------------
* Fixed leak when closing a connection, LM-59
* Escape the resource string (fixes problems with iChat contacts), LM-48
* Stale incoming source destroyed/unreffed, LM-60
* Fixed a leak in the synchronous layer, LM-66
* Fixed a couple of warning when building on 64 bit alpha, LM-41.

Changes in 1.1.4:
-----------------
* Generate better message IDs, LM-56
* Fixed a leaking proxy object, LM-51
* Fixed random bugs on connection errors, LM-58
* Removed a warning when using proxy, LM-50.

Changes in 1.1.3:
-----------------
* Infinite reading of 0 bytes when disconnected using gnutls, LM-45
* Loudmouth exports some internal SHA* functions, LM-47
* Fix auto* warning, LM-46

Changes in 1.1.2:
-----------------
* Support for internationalised domain names, LM-42
* Fixed issue where no disconnect signal was emitted when connection closed
  during opening, LM-43
* Plugged a memory leak in LmConnection, LM-39
* Fixed warnings due to uninitialized value, LM-35
* Fixed a reentrancy-related issue with LmConnection, LM-37

Changes in 1.0.3:
-----------------
* Quick build fix release when building without SSL, LM-25.

Changes in 1.0.2:
-----------------
* Fixed issue when opening in blocking mode failed, LM-15
* Unregister sources when connection fails to avoid later segfault, LM-16
* Fixed an issue where large messages would get trunkated, LM-17

Changes in 1.0.1:
-----------------
* Unref contexts when freeing the connection          (Mikael Hallendal)

Changes in 1.0:
---------------
* Fix crash when connection goes down using SSL, LM-3 (Mikael Hallendal)
* Fixed some GCC 4.0 warnings                         (Mikael Hallendal)
* Some API fixups for consistency in LmSLL            (Mikael Hallendal)
* Don't try to setup SSL if it's not supported        (Martyn Russel)
* Documentation fixes                                 (Mikael Hallendal)
* State fixes when using blocking API, LM-7           (Mikael Hallendal)
* Check that connection is open before using it, LM-6 (Mikael Hallendal)
* Fix infinite loop in blocking open if failed, LM-8  (Mikael Hallendal)

Changes in 0.90:
----------------
* Two crashers for ssl read fixed (Micke, Daniel Lavalliere)
* Fixed a couple of leaks when freeing the connection (Daniel Lavalliere)
* A couple of thread related issues solved (Micke, Daniel Lavalliere)
* Work with latest GnuTLS (foser)
* Improved the example tools some (Micke)

Changes in 0.17.2:
------------------
* Unset proxy by setting it to NULL.
* Fixes for 64 bit architectures.
* Fixed an issue when not using default main context.
* Fixed UTF-8 issue that lead to a few weird problems in Gossip.

Changes in 0.17.1:
------------------
* Fixed problem when not setting a keep alive rate.

Changes in 0.17:
----------------
* Compile fixes (Sjoerd Simons
* Depends on GLib 2.4
* Fixed memory leak
* Added support to work tunnel, using another JID than the connect server.
* Support for keep alive packages

Changes in 0.16:
----------------
* Support for giving the context to which an LmConnection should run in
  (should solve issues when wanting to run several LmConnections in various
   threads)
* HTTP proxy support
* Memory leak fix
* Distribute the win32-config.h file needed to build on Windows.

Changes in 0.15.1:
------------------
* Require GnuTLS >= 1.0.0

Changes in 0.15:
----------------
* Better Win32 instructions.
* Fixed bug when disconnecting and then connecting again using the same
  LmConnection.
* Documentation typos and errors fixed by Ross Burton.
* Better SSL support, now gives the application the possibility to decide
  if it wants to connect on bogus SSL certificates.
* The SSL API changed a bit, take a look at the API docs.
* Fixed timestamp error on FreeBSD.

Changes in 0.14.1:
------------------
* Patch from foser@gentoo.org for a problem when using GnuTLS.

Changes in 0.14:
----------------
* Added information about building on Windows
* Bug when using GnuTLS fixed.
* Fixed a bug that could cause replies to be overlooked when using
  blocking mode.
* Added IPv6-support, patch from Sjoerd Simons.
* Made it easier to build RPMs without SSL-support.
* Fixed a memleak when destroying a connection.

Changes in 0.13.2:
------------------
* Don't crash when we receive </stream:stream>

Changes in 0.13.1:
------------------
* Don't require GnuTLS to build

Changes in 0.13:
----------------
* Fixed bug in the sub type checking code for messages
* Fixed bug that hang the app when disconnected uncleanly
* Fixed autoconf problems with gnutls
* Disconnect cleanly
* Fix sha implementation to work on Linux/PPC

Changes in 0.12:
----------------
* Added support for cancelling messages waiting for reply
* Now checks for supported auth methods.
* Fixed RPM's to require GnuTLS.

Changes in 0.11:
----------------
* SSL-support, patch by Bartosz Zapalowski

Changes in 0.10.1:
------------------
* Build fix for *BSD. Reported by Ralph Meijer.

Changes in 0.10:
----------------
* Fixed case bug in message nodes. Making it not treat the XML as case
  dependant.
* Added lm_message_node_[get|set]_raw_mode that determines if
  the node value should be escaped or not.
* Return correct type from lm_message_get_sub_type.
* Win32 build fixes. Sent in by Vince Negri.

Changes in 0.9:
---------------
* Build fixes Solaris
* Non async operations
* API additions to ease binding efforts
* Added Debug output support
* Bug fixes
* API changes (see below).
* Support for disconnect and disconnect signalling.


examples/test-lm.c
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * Copyright (C) 2003-2006 Imendio AB
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <config.h>

#include <string.h>
#include <stdlib.h>

#include <glib.h>
#include <loudmouth/loudmouth.h>

static GMainLoop *main_loop = NULL;
static gboolean   test_success = FALSE;

static gchar      expected_fingerprint[20];

static gchar     *server = NULL;
static gint       port = 5222;
static gchar     *username = NULL;
static gchar     *password = NULL;
static gchar     *resource = "test-lm";
static gchar     *fingerprint = NULL;

static GOptionEntry entries[] =
{
    { "server", 's', 0, G_OPTION_ARG_STRING, &server,
      "Server to connect to", NULL },
    { "port", 'P', 0, G_OPTION_ARG_INT, &port,
      "Port to connect to [default=5222]", NULL },
    { "username", 'u', 0, G_OPTION_ARG_STRING, &username,
      "Username to connect with (user@server.org)", NULL },
    { "password", 'p', 0, G_OPTION_ARG_STRING, &password,
      "Password to try", NULL },
    { "resource", 'r', 0, G_OPTION_ARG_STRING, &resource,
      "Resource connect with [default=test-lm]", NULL },
    { "fingerprint", 'f', 0, G_OPTION_ARG_STRING, &fingerprint,
      "SSL Fingerprint to use", NULL },
    { NULL }
};

static gchar *
get_part_name (const gchar *username)
{
    const gchar *ch;

    g_return_val_if_fail (username != NULL, NULL);

    ch = strchr (username, '@');
    if (!ch) {
        return NULL;
    }

    return g_strndup (username, ch - username);
}

static void
print_finger (const char   *fpr,
              unsigned int  size)
{
    gint i;
    for (i = 0; i < size-1; i++) {
        g_printerr ("%02X:", fpr[i]);
    }

    g_printerr ("%02X", fpr[size-1]);
}

static LmSSLResponse
ssl_cb (LmSSL       *ssl,
        LmSSLStatus  status,
        gpointer     ud)
{
    g_print ("TestLM: SSL status:%d\n", status);

    switch (status) {
    case LM_SSL_STATUS_NO_CERT_FOUND:
        g_printerr ("TestLM: No certificate found!\n");
        break;
    case LM_SSL_STATUS_UNTRUSTED_CERT:
        g_printerr ("TestLM: Certificate is not trusted!\n");
        break;
    case LM_SSL_STATUS_CERT_EXPIRED:
        g_printerr ("TestLM: Certificate has expired!\n");
        break;
    case LM_SSL_STATUS_CERT_NOT_ACTIVATED:
        g_printerr ("TestLM: Certificate has not been activated!\n");
        break;
    case LM_SSL_STATUS_CERT_HOSTNAME_MISMATCH:
        g_printerr ("TestLM: Certificate hostname does not match expected hostname!\n");
        break;
    case LM_SSL_STATUS_CERT_FINGERPRINT_MISMATCH: {
        const char *fpr = lm_ssl_get_fingerprint (ssl);
        g_printerr ("TestLM: Certificate fingerprint does not match expected fingerprint!\n");
        g_printerr ("TestLM: Remote fingerprint: ");
        print_finger (fpr, 16);

        g_printerr ("\n"
                    "TestLM: Expected fingerprint: ");
        print_finger (expected_fingerprint, 16);
        g_printerr ("\n");
        break;
    }
    case LM_SSL_STATUS_GENERIC_ERROR:
        g_printerr ("TestLM: Generic SSL error!\n");
        break;
    }

    return LM_SSL_RESPONSE_CONTINUE;
}

static void
connection_auth_cb (LmConnection *connection,
                    gboolean      success,
                    gpointer      user_data)
{
    if (success) {
        LmMessage *m;

        test_success = TRUE;
        g_print ("TestLM: Authenticated successfully\n");

        m = lm_message_new_with_sub_type (NULL,
                                          LM_MESSAGE_TYPE_PRESENCE,
                                          LM_MESSAGE_SUB_TYPE_AVAILABLE);
        lm_connection_send (connection, m, NULL);
        g_print ("TestLM: Sent presence message:'%s'\n",
                 lm_message_node_to_string (m->node));

        lm_message_unref (m);
    } else {
        g_printerr ("TestLM: Failed to authenticate\n");
        g_main_loop_quit (main_loop);
    }
}

static void
connection_open_cb (LmConnection *connection,
                    gboolean      success,
                    gpointer      user_data)
{
    if (success) {
        gchar *user;

        user = get_part_name (username);
        lm_connection_authenticate (connection, user,
                                    password, resource,
                                    connection_auth_cb,
                                    NULL, FALSE,  NULL);
        g_free (user);

        g_print ("TestLM: Sent authentication message\n");
    } else {
        g_printerr ("TestLM: Failed to connect\n");
        g_main_loop_quit (main_loop);
    }
}

static void
connection_close_cb (LmConnection       *connection,
                     LmDisconnectReason  reason,
                     gpointer            user_data)
{
    const char *str;

    switch (reason) {
    case LM_DISCONNECT_REASON_OK:
        str = "LM_DISCONNECT_REASON_OK";
        break;
    case LM_DISCONNECT_REASON_PING_TIME_OUT:
        str = "LM_DISCONNECT_REASON_PING_TIME_OUT";
        break;
    case LM_DISCONNECT_REASON_HUP:
        str = "LM_DISCONNECT_REASON_HUP";
        break;
    case LM_DISCONNECT_REASON_ERROR:
        str = "LM_DISCONNECT_REASON_ERROR";
        break;
    case LM_DISCONNECT_REASON_UNKNOWN:
    default:
        str = "LM_DISCONNECT_REASON_UNKNOWN";
        break;
    }

    g_print ("TestLM: Disconnected, reason:%d->'%s'\n", reason, str);
}

static LmHandlerResult
handle_messages (LmMessageHandler *handler,
                 LmConnection     *connection,
                 LmMessage        *m,
                 gpointer          user_data)
{
    g_print ("TestLM: Incoming message from: %s\n",
             lm_message_node_get_attribute (m->node, "from"));

    return LM_HANDLER_RESULT_REMOVE_MESSAGE;
}

int
main (int argc, char **argv)
{
    GOptionContext   *context;
    LmConnection     *connection;
    LmMessageHandler *handler;
    gboolean          result;
    GError           *error = NULL;

    context = g_option_context_new ("- test Loudmouth");
    g_option_context_add_main_entries (context, entries, NULL);
    g_option_context_parse (context, &argc, &argv, NULL);
    g_option_context_free (context);

    if (!server || !username || !password) {
        g_printerr ("For usage, try %s --help\n", argv[0]);
        return EXIT_FAILURE;
    }

    if (fingerprint && !lm_ssl_is_supported ()) {
        g_printerr ("TestLM: SSL is not supported in this build\n");
        return EXIT_FAILURE;
    }

    if (username && strchr (username, '@') == NULL) {
        g_printerr ("TestLM: Username must have an '@' included\n");
        return EXIT_FAILURE;
    }

    connection = lm_connection_new (server);
    lm_connection_set_port (connection, port);
    lm_connection_set_jid (connection, username);

    handler = lm_message_handler_new (handle_messages, NULL, NULL);
    lm_connection_register_message_handler (connection, handler,
                                            LM_MESSAGE_TYPE_MESSAGE,
                                            LM_HANDLER_PRIORITY_NORMAL);

    lm_message_handler_unref (handler);

    lm_connection_set_disconnect_function (connection,
                                           connection_close_cb,
                                           NULL, NULL);

    if (fingerprint) {
        LmSSL *ssl;
        char  *p;
        int    i;

        if (port == LM_CONNECTION_DEFAULT_PORT) {
            lm_connection_set_port (connection,
                                    LM_CONNECTION_DEFAULT_PORT_SSL);
        }

        for (i = 0, p = fingerprint; *p && *(p+1); i++, p += 3) {
            expected_fingerprint[i] = (unsigned char) g_ascii_strtoull (p, NULL, 16);
        }

        ssl = lm_ssl_new (expected_fingerprint,
                          (LmSSLFunction) ssl_cb,
                          NULL, NULL);

        lm_ssl_use_starttls (ssl, TRUE, FALSE);

        lm_connection_set_ssl (connection, ssl);
        lm_ssl_unref (ssl);
    }

    result = lm_connection_open (connection,
                                 (LmResultFunction) connection_open_cb,
                                 NULL, NULL, &error);

    if (!result) {
        g_printerr ("TestLM: Opening connection failed, error:%d->'%s'\n",
                    error->code, error->message);
        g_free (error);
        return EXIT_FAILURE;
    }

    main_loop = g_main_loop_new (NULL, FALSE);
    g_main_loop_run (main_loop);

    return (test_success ? EXIT_SUCCESS : EXIT_FAILURE);
}
